src/hotspot/share/utilities/exceptions.cpp

```diff
@@ -262,8 +263,24 @@ void Exceptions::fthrow(JavaThread* thread, const char* file, int line, Symbol*
   va_list ap;
   va_start(ap, format);
   char msg[max_msg_size];
-  os::vsnprintf(msg, max_msg_size, format, ap);
+  int ret = os::vsnprintf(msg, max_msg_size, format, ap);
   va_end(ap);
+  // If ret == -1 then either there was a format conversion error, or the required buffer size
+  // exceeds INT_MAX and so couldn't be returned (undocumented behaviour of vsnprintf). Depending
+  // on the platform the buffer may be filled to its capacity (Linux), filled to the conversion
+  // that encountered the overflow (macOS), or is empty (Windows), so it is possible we
+  // have a truncated UTF-8 sequence. Similarly, if the buffer was too small and ret >= max_msg_size
+  // we may also have a truncated UTF-8 sequence. In such cases we need to fix the buffer so the UTF-8
+  // sequence is valid.
+  if ((ret == -1 || ret >= max_msg_size) && strlen(msg) > 0) {
```

Comment:
Do we need to check if `strlen(msg) == max_msg_size - 1`? If strlen is shorter, the bytes between the null terminator and max_msg_size are undefined, which might trigger an assertion while truncating.

Reply:
In fact we know it may be shorter than `max_msg_size - 1` - that is what we get on macOS if the string is huge and exceeds `INT_MAX` causing `vsnprintf` to return -1. I originally had an assert that failed due to that.

I need to fix this case as well. \<sigh>. Good catch.

Reply:
I did some experimentation here and it seems in practice that if the buffer is only partially filled then it will contain valid data as `vsnprintf` would stop filling at a well-defined point. But as it is not a clearly specified area we pass the buffer through anyway, using `strlen(msg)` . Most of the time a partially filled buffer will end with an ASCII character anyway and so the utf8 truncation operation will be a no-op.